lib/semantics/resolve-names.cc

```diff
@@ -797,6 +804,14 @@ class ConstructVisitor : public DeclarationVisitor {
   void Post(const parser::ExitStmt &x) { CheckRef(x.v); }
 
 private:
+  // The represents: associate-name => expr | variable
```

Comment:
I preserved the variable when it is present because I think it is needed when its type is `CLASS(t)`. For example, see the association `c => z` in `symbol11.f90`. When `z` is evaluated as an expression, its DynamicType is `TYPE(t)`. But within the ASSOCIATE construct I think the type of `c` needs to be `CLASS(t)`. Another instance of this is in the CLASS DEFAULT in `s3`.

If the original variable is not need for this case I can probably eliminate it in `association_`.